src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -22,6 +25,19 @@ internal CoseSign1Message(CoseHeaderMap protectedHeader, CoseHeaderMap unprotect
             _signature = signature;
         }
 
+        /// <summary>
+        /// Signs the specified content and encodes it as a COSE_Sign1 message with detached content.
+        /// </summary>
+        /// <param name="detachedContent">The content to sign.</param>
+        /// <param name="signer">The signer information used to sign <paramref name="detachedContent"/>.</param>
+        /// <param name="associatedData">The extra data associated with the signature, which must also be provided during verification.</param>
+        /// <returns>The encoded message.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="detachedContent"/> is <see langword="null"/>.
+        /// -or-
+        /// <paramref name="signer"/> is <see langword="null"/>.</exception>
+        /// <exception cref="CryptographicException">Protected and unprotected buckets contain duplicate labels.
+        /// -or-
+        /// One of the specified labels in <see cref="CoseHeaderLabel.CriticalHeaders"/> is missing.</exception>
```

Comment:
These could as well be `ArgumentException`, but consider that we do the same check on `Decode*` and `*Encode`, in both of those cases it can't be `ArgumentException`.

Reply:
Right.  Same "bad thing", but three different reasons:

* Sign: You gave me the values, you could have avoided the duplicate: ArgumentException.
* Encode: You added value to UnprotectedHeaders that was already in ProtectedHeaders.  Encode can no longer work.  This is state, not an argument, so InvalidOperationException.
* Decode: I'm refusing to decode this value because it isn't spec-compliant.  There's nothing you can do about this, except complain to the person who wrote it.  It's sort of an ArgumentException, except you couldn't have checked it yourself.  It's not state, so it's not InvalidOperationException.  So it's "some other exception", and in the case of crypto libraries, that's CryptographicException.